### DISCUSSION TITLE
How do interface routes defined in Nextjs 13 interact with UI components in a friendly way ?

### DISCUSSION THREAD

**Discussion by evanlong-me** [vercel/next.js#55701] [Help] [answered]
### Summary

I wonder what the solution is. 🤨

### Additional information

```js
// app/api/pay/notify/route.ts
import { NextResponse } from "next/server"

export async function POST(req: Request) {
  const data = await req.json()
  // Interact with UI components
  // ......
}
```

**Comment by icyJoseph** [comment]
These API handlers happen in a separate, let's say, JavaScript Context. You have the client side context, totally isolated from the server, and within the server you have each handler running without knowledge of each other.

So when the server receives a client request, with method POST, this handler is activated. There's no JavaScript to interact with the React component tree. There's also no way, AFAIK, for these handlers to reach the React Tree. 

Rather, your best chance, again AFAIK, is to fly a request from the React tree, from event handlers, effects, etc, to this API route.

What's more, the above is not really feasible, in a sensible way, during build time.

May I ask what is it you are trying to do?

> In another framework, Nuxt, which uses the Nitro engine, there's something called Direct API calls, https://nuxt.com/docs/guide/concepts/server-engine#direct-api-calls but that's again, from the client to the server route handler

**Reply by evanlong-me** [reply]
In fact, this code example is a payment notification interface. I want to give some friendly prompts after the user successfully pays.

**Reply by evanlong-me** [reply]
This interface receives some JSON information about payment completion. I thought we could use it as a cue. If not, I can only do the redirect page.

**Reply by icyJoseph** [reply, marked as answer]
I think because it is HTTP, maybe you could send down events, as things happen, https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events, or a very short-lived websocket, but other than that 🤔 I don't think so

Another way would be to architect things differently, and batch the payment job into a queue, and give the client an ID, which then can be used to poll, for the status of the payment job in the queue. 

But something like, work with the UI directly from the API route, I don't know about.

**Reply by evanlong-me** [reply]
You mentioned polling, which seems to remind me that polling seems to solve my problem, thank you 😊 ~

**Reply by icyJoseph** [reply]
Hehe awesome! Please don't forget to mark as answered :)
